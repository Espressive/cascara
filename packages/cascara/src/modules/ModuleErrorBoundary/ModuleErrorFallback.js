import React from 'react';

import styles from '../DataModule.module.scss';

const ModuleErrorFallback = () => {
  return (
    <div className={styles.Error} role='alert'>
      Error
    </div>
  );
};

export default ModuleErrorFallback;
