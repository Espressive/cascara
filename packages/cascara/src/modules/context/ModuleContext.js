import { createContext } from 'react';

// We should ONLY put values here that are needed for ALL Modules
const defaultValue = {
  isEditing: false,
  isEditingAny: false,
  value: 'blue',
};

/*
 * We will use this baseline context to create any of our context providers from.
 * Any unique functionality should be included in that provider and not in this Context.
 */
const ModuleContext = createContext(defaultValue);

/*
 * We are also exporting the default value in case we need to hydrate or merge
 * any providers with these default values.
 */
export { defaultValue };

export default ModuleContext;
