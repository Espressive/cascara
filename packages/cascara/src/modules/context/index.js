export { default as ModuleContext } from './ModuleContext';
export { default as ModuleProvider } from './ModuleProvider';
