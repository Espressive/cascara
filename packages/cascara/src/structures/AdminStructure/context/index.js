export { default as AdminContext } from './AdminContext';
export { default as AdminProvider } from './AdminProvider';
