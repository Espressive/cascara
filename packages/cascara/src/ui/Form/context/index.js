import FormProvider from './FormProvider';

export { FormProvider };
