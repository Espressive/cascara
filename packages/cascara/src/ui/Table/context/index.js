export * from './TableContext';
export { default } from './TableContext';
