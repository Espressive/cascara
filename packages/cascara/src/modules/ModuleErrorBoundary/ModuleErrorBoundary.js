import React from 'react';
import pt from 'prop-types';

import { Boundaries } from '../../system-components';
import ModuleErrorFallback from './ModuleErrorFallback';

const propTypes = {
  children: pt.oneOfType([pt.arrayOf(pt.node), pt.node]),
};

const ModuleErrorBoundary = ({ children }) => {
  return (
    <Boundaries ErrorComponent={ModuleErrorFallback}>{children}</Boundaries>
  );
};

ModuleErrorBoundary.propTypes = propTypes;

export default ModuleErrorBoundary;
