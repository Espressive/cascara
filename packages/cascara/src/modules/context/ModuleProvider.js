import React from 'react';
import pt from 'prop-types';

import ModuleContext, { defaultValue } from './ModuleContext';
import useToggle from '../../hooks/useToggle';

const propTypes = {
  children: pt.oneOfType([pt.element, pt.arrayOf(pt.element)]),
  value: pt.shape({
    isEditing: pt.bool,
  }),
};

const ModuleProvider = ({ children, value, ...props }) => {
  const [isEditing, setIsEditing] = useToggle(value?.isEditing || false);

  const mergedValues = {
    ...defaultValue,
    isEditing,
    setIsEditing,
    ...value,
  };

  return (
    <ModuleContext.Provider value={mergedValues} {...props}>
      {children}
    </ModuleContext.Provider>
  );
};

ModuleProvider.propTypes = propTypes;

export default ModuleProvider;
