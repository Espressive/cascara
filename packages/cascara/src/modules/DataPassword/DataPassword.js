import React, { useContext } from 'react';
import { Input } from 'reakit/Input';
import pt from 'prop-types';
import { ModuleContext } from '../context';
import styles from '../DataModule.module.scss';

import ModuleErrorBoundary from '../ModuleErrorBoundary';

const propTypes = {
  /** A module can have an Attribute, which will be used as form field name */
  attribute: pt.string,
  /** A Module can be defined to not present an editing state */
  isEditable: pt.bool,
  /** Presents the input without a label. NOT USER CONFIGURABLE */
  isLabeled: pt.bool,
  /** A Module needs to have a unique label relative to its context */
  label: pt.string,
  /** A Module can have a value */
  value: pt.string,
};

const DataPassword = ({
  attribute,
  isEditable = true,
  isLabeled = true,
  label,
  value,
  ...rest
}) => {
  const { isEditing, formMethods } = useContext(ModuleContext);

  // NOTE: THESE TWO SET DEFINITIONS COULD PROBABLY BECOME A HELPER FUNCTION FOR USE IN ALL MODULES
  // We do not want to add a redundant aria-label property if there
  // is an html label present with a linking `for` attribute.
  const setAriaLabel = isLabeled ? undefined : label;
  // We do not want to set a for attribute if there is no label content
  // because we are defining aria label instead
  const setHtmlFor = isLabeled ? label : undefined;

  const renderEditing = (
    <label htmlFor={setHtmlFor}>
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
      <Input
        {...rest}
        aria-label={setAriaLabel}
        className={styles.Input}
        defaultValue={value}
        id={label}
        name={attribute || label}
        ref={formMethods?.register}
        type='password'
      />
    </label>
  );

  const renderDisplay = (
    <span>
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
      <span aria-label={label} className={styles.Input} {...rest}>
        ******
      </span>
    </span>
  );

  // Do not render an editable input if the module is not editable
  return (
    <ModuleErrorBoundary>
      <div className={styles.Text}>
        {isEditing && isEditable ? renderEditing : renderDisplay}
      </div>
    </ModuleErrorBoundary>
  );
};

DataPassword.propTypes = propTypes;

export { propTypes };
export default DataPassword;
