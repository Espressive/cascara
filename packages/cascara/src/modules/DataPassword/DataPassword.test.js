import { fireEvent, render, screen } from '@testing-library/react';

import cosmosFixtures, {
  displayProps,
  editingProps,
} from './DataPassword.fixture';

// We cannot destructure during import because the default export in Cosmos
// multi-fixture files is an object so we need to import the fixtures first,
// then destructure them separately.
const { display, editing, displayNoLabel, editingNoLabel } = cosmosFixtures;

describe('DataPassword', () => {
  // without ModuleSandbox will render the property information into a span
  describe('display', () => {
    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(display).container;
    });

    test('snapshot', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <span> by default', () => {
      const input = screen.getByLabelText(displayProps.label);
      // Make sure the actual DOM element is not render an input
      expect(input.tagName).toMatch('SPAN');
      // Make sure the dom element that has our aria-label is the input
      expect(input.classList.contains('Password')).toBe(true);
    });
  });

  describe('editing', () => {
    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(editing).container;
    });

    test('snapshot', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <input date> by default', () => {
      const input = screen.getByLabelText(editingProps.label);
      // Check that we also use the correct type
      expect(input).toHaveAttribute('type', 'password');
      expect(input).toHaveValue('mypassword');
    });

    test('change value', () => {
      const newPassword = 'newPassword';
      const input = screen.getByLabelText(editingProps.label);
      fireEvent.change(input, { target: { value: newPassword } });
      expect(input).toHaveValue(newPassword);
    });
  });

  describe('accessibility', () => {
    test('editing', () => {
      render(editing);

      const input = screen.getByLabelText(editingProps.label);
      // The label tag is the parent wrapper
      const label = input.closest('label');

      // Test is written this way to make sure we know that both values need to be the same.
      const linkedLabelValue = editingProps.label;

      // Verify label for attribute has linked value
      expect(label).toHaveAttribute(
        'for',
        expect.stringContaining(linkedLabelValue)
      );
      // Verify input id attribute has linked value
      expect(input).toHaveAttribute(
        'id',
        expect.stringContaining(linkedLabelValue)
      );
      // Check that the input does NOT have an aria-label defined because there is a label tag
      expect(input).not.toHaveAttribute('aria-label');
    });

    test('display no label', () => {
      // Make sure that the input is still accessible with label text even when we are not showing a label tag in tables
      render(displayNoLabel);
      const input = screen.getByLabelText(displayProps.label);
      expect(input).toBeDefined();
    });

    test('editing no label', () => {
      // Make sure that the input is still accessible with label text even when we are not showing a label tag in tables
      render(editingNoLabel);
      const input = screen.getByLabelText(editingProps.label);
      expect(input).toBeDefined();
    });
  });
});
