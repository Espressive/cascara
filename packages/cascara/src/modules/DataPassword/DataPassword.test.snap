// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataPassword display snapshot 1`] = `
<div>
  <div
    class="Text"
  >
    <span>
      <span
        class="LabelText"
      >
        Display
      </span>
      <span
        aria-label="Display"
        class="Password"
      >
        ******
      </span>
    </span>
  </div>
</div>
`;

exports[`DataPassword editing snapshot 1`] = `
<div>
  <div
    class="Text"
  >
    <label
      for="Editing"
    >
      <span
        class="LabelText"
      >
        Editing
      </span>
      <input
        class="Password"
        id="Editing"
        name="Editing"
        type="password"
        value="mypassword"
      />
    </label>
  </div>
</div>
`;
